File: frontend/src/api/chatbot.ts
const BASE = `${(import.meta.env.VITE_API_URL ?? "").replace(/\/$/, "")}/api`;

export interface ChatMessage {
  role: "user" | "assistant";
  content: string;
}

export interface SourceDocument {
  filename: string;
  chunk_preview: string;
}

export interface ChatMeta {
  used_fallback: boolean;
  response_basis: "documented" | "partial_document" | "general_knowledge";
  sources: SourceDocument[];
}

export async function sendChatMessage(
  message: string,
  conversationHistory: ChatMessage[],
  token: string,
  babyId?: string,
  onChunk?: (text: string) => void,
  onMeta?: (meta: ChatMeta) => void,
  signal?: AbortSignal,
): Promise<void> {
  const response = await fetch(`${BASE}/chatbot/chat`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({
      message,
      conversation_history: conversationHistory,
      baby_id: babyId ?? null,
    }),
    signal,
  });

  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error((body as any).detail || "오류가 발생했습니다.");
  }

  const reader = response.body!.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n\n");
    buffer = lines.pop() ?? "";

    for (const line of lines) {
      if (!line.startsWith("data: ")) continue;
      const raw = line.slice(6).trim();
      if (!raw) continue;

      const event = JSON.parse(raw) as { type: string; [key: string]: unknown };

      if (event.type === "meta") {
        onMeta?.(event as unknown as ChatMeta);
      } else if (event.type === "chunk") {
        onChunk?.(event.text as string);
      } else if (event.type === "error") {
        throw new Error((event.message as string) || "오류가 발생했습니다.");
      }
    }
  }
}

File: backend/app/schemas/chatbot.py
from __future__ import annotations

import uuid
from typing import Literal

from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    role: str  # "user" | "assistant"
    content: str


class ChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=2000, description="사용자 메시지")
    conversation_history: list[ChatMessage] = Field(
        default_factory=list,
        description="이전 대화 기록 (최근 순서로 전달)",
    )
    baby_id: uuid.UUID | None = Field(None, description="현재 선택된 아기 프로필 ID")


class SourceDocument(BaseModel):
    filename: str
    chunk_preview: str


class ChatResponse(BaseModel):
    answer: str
    used_fallback: bool = Field(description="True = 지식 베이스 미발견, 모델 자체 지식 사용")
    response_basis: Literal["documented", "partial_document", "general_knowledge"] = Field(
        description="documented=문서 기반, partial_document=문서 일부 근거/구체 정보 부족, general_knowledge=일반 지식 포함",
    )
    sources: list[SourceDocument] = Field(default_factory=list)
